### DISCUSSION TITLE
Extrat Data from Multiple Tables in Serenity

### DISCUSSION THREAD

**Discussion by eworkonlinegroup** [serenity-is/Serenity#6658] [Help]
I have 3 Related Tables in Sql Server 

CREATE TABLE [dbo].[Purchases](
	[Id] [bigint] IDENTITY(1,1) NOT NULL,	
	[BillOfMaterialVersionId] [bigint] NULL,
	[InvoiceNo] [nvarchar](250) NULL,
	[ChallanNo] [nvarchar](250) NULL,	
 CONSTRAINT [PK_Purchases] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO

CREATE TABLE [dbo].[BillOfMaterialVersion](
	[Id] [bigint] IDENTITY(1,1) NOT NULL,
	[BOMVersion] [nvarchar](50) NULL,	
 CONSTRAINT [PK_BillOfMaterialVersion] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO

CREATE TABLE [dbo].[BillOfMaterial](
	[Id] [bigint] IDENTITY(1,1) NOT NULL,	
	[BillOfMaterialVersionId] [bigint] NULL,
	[RawMaterialName] [nvarchar](250) NULL,	
 CONSTRAINT [PK_BillOfMaterial] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO

I wish to show Data from all 3 tables in a  Report and my base table of report is Purchases Table / Row

I am able to extract data of BillOfMaterialVersion through PurchasesRow Using LeftJoin and Foreign Key Attributes, as Purchase Table as  Foreign Key field of BillOfMaterialVersion Table via BillOfMaterialVersionId 

But How Could I Extract Data of BillOfMaterial Table as well, in same report?
As BillOfMaterialVersion Table doesn't have Foreign Key of BillOfMaterial Table, infact BillOfMaterial Table have a Foreign Key of BillOfMaterialVersion Table via BillOfMaterialVersionId.

**Comment by luongnv6** [comment]
try use Store procedue, this is demo
https://github.com/serenity-is/common-features/blob/master/src/Serenity.Demo.Northwind/Modules/Reports/CustomerGrossSalesReport.cs

**Comment by muhammedyaman** [comment]
I find it the easiest way to create a view with all the data I need and then use that view for reports. I wrote something like this from what I understand:
```
SELECT 
	 P.Id PurchaseId
        ,P.InvoiceNo 
	,P.ChallanNo
	,BOMV.Id BOMVersionId
	,BOMV.BOMVersion BOMVersion
	,BOM.Id BOMId
	,BOM.RawMaterialName BOMRawMaterialName
FROM Purchases P
LEFT JOIN BillOfMaterialVersion BOMV ON BOMV.Id = P.BillOfMaterialVersionId
LEFT JOIN BillOfMaterial BOM ON BOM.BillOfMaterialVersionId = BOMV.Id
```
And the output will be: 
![image](https://user-images.githubusercontent.com/44383772/214940090-b29c784f-4510-4649-8dba-cb9c10e3f68d.png)
